lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,55 +26,95 @@ def __init__(self, context):
+        else:
             # We were not given any remotes on the command line, build the list
             # from the tools group directory.
             remotes = self.remote(self.tool_group_dir)
-            if not remotes:
-                # FIXME:  This is a weird case -- the group directory exists,
-                #  but contains no remotes. This is the result of clearing the
-                #  tools from a group which has already been cleared, because
-                #  the first clear does not remove the group directory.  This
-                #  behavior might be desirable for the `default` directory, but
-                #  it seems wrong for custom groups.
-                self.logger.error(f'No such group "{self.context.group}".')
-                return 1
-
-        tool_not_found = True  # Not found, yet
+            if (
+                not remotes
+                and group != "default"
+                and self.is_empty(self.tool_group_dir)
+            ):
+                # Remove non-default empty tool directories
+                try:
+                    shutil.rmtree(self.tool_group_dir)
+                except OSError as e:
+                    self.logger.error(
+                        "Failed to remove empty group directory %s\n%s",
+                        self.tool_group_dir,
+                        str(e),
+                    )
+
         for remote in remotes:
+            tools_not_found_remote = []
             tg_dir_r = self.tool_group_dir / remote
             if not tg_dir_r.exists():
                 self.logger.warn(
-                    'No remote host "%s" in group %s.', remote, self.context.group,
+                    'No remote host "%s" in group %s.', remote, group,
                 )
                 continue
 
             if self.context.name:
-                names = [self.context.name]
+                names = self.context.name.split(",")
             else:
                 # Discover all the tools registered for this remote
                 names = self.tools(tg_dir_r)
                 if not names:
                     # FIXME:  this is another odd case -- the remote subdirectory
                     #  exists, but it's empty.  (We'll remove it below.)
                     self.logger.warn(
-                        'No tools in group "%s" on host "%s".',
-                        self.context.group,
-                        remote,
+                        'No tools in group "%s" on host "%s".', group, remote,
                     )
 
             for name in names:
-                status = self._clear_tools(name, remote)
-                if status > 0:
-                    errors = 1
-                elif status == 0:
-                    tool_not_found = False
+                status = self._clear_tools(name, remote, group)
+                if status:
+                    tools_not_found_remote.append(name)
+                    if status > 0:
+                        errors = 1
```

Comment:
This logic looks a bit awkward. I think it's because within a few lines you're treating `status` both as a boolean and an integer. And it's perfectly legal Python: if not successful, append the tool, but count it as an error only if the tool was found. This however forces the reader to go look at `_clear_tools` to figure out why `status` might be truthy but not `> 0`.

I think it might be a bit more clear to use `if status != 0` and keep it all in the integer family. (Although I'm not thrilled about the whole original `-1` for missing tool, `1` for failure removing thing in the first place, that's not new code.)

Reply:
I still think the combination of boolean and integer usages of `status` here is unnecessary and unclear; however fixing that still wouldn't explain the weird tri-state return from `_clear_tools` and I'm not crazy about capturing it in a comment here. In an IDE that'll be handled by a popup on the name, so maybe I'm just complaining too much. 🤐